File: linkedlist/double_linked_list/DoubleLinkedList.py
"""  Create a link list and add items to the head or middle or end of the Linkedlist.

 * Double linked list in Java
 * Insertion:  O(1)
 * Search: O(1)
 * Deletion: O(1)
 * @author tajayebi
 """

class Node(object):
    #__slots__ = ['data','next_ptr','prev_ptr']  # Tell python not to use dictionary for instance variables
    def __init__(self, data: object, next_node: object = None, prev_node: object = None) -> object:
        self.data = data
        self.next_ptr = next_node
        self.prev_ptr = prev_node

    def getData(self):
        return self.data

    def setData(self, data):
        self.data = data

    def getNextNode(self):
        return self.next_ptr

    def setNextNode(self, Node):
        self.next_ptr = Node

    def getPrevNode(self):
        return self.prev_ptr

    def setPrevNod(self, Node):
        self.prev_ptr = Node


class LinkList:
    """ Lx$inkedList class constructor, vaiables with underscore are private class variables """

    def __init__(self, head=None, tail=None):
        self._size = 0
        self._data = 0
        self._head = head
        self._tail = tail

    def getSize(self):
        return self._size

    def incrementSize(self):
        self._size += 1

    def isEmpty(self):
        """ check to see if the list empty """
        return self._size <= 0

    def addNode(self, newValue):
        newNode = Node(newValue)

        currentNode = self._head
        # if LinkedList is empty
        if (self.getSize() <= 0):
            self._head = newNode
            self._tail = newNode
            self.incrementSize()
            print( newNode.data," Added first item")
            return True

        while currentNode is not None:
              if newNode.data <= currentNode.data:
                  if currentNode.prev_ptr is None:
                      newNode.next_ptr = self._head
                      self._head.previous_ptr = newNode
                      print (newNode.data," Added to Head")
                      self._head = newNode;
                      self.incrementSize()
                      return True

                  if  currentNode.prev_ptr is not None:

                      currentNode.prev_ptr = newNode // bug

                      newNode.next_ptr = currentNode
                      newNode.prev_ptr = currentNode.prev_ptr
                      self.incrementSize()
                      print(newNode.data , " Added to middle")
                      return True
              else:
                  if currentNode.next_ptr is None:
                      currentNode.next_ptr= newNode
                      newNode.previous_ptr = currentNode
                      self._tail = newNode
                      self.incrementSize()
                      print(newNode.data , " Added to tail")
                      return True

                  else:
                      currentNode = currentNode.next_ptr



    def removeNode(self, node_value):
        previousNode = None
        currentNode = self.head
        while currentNode is not None:
            if currentNode.getData() == node_value:
                if previousNode:
                    previousNode.setNextNode(currentNode.getNextNode())
                else:
                    self.head = currentNode.getNextNode()
                return True
            previouNode = currentNode
            currentNode = currentNode.getNextNode()
            return False

    def printList(self):
        currentNode = self._head
        print('Total number of items in the LinkedList:', self.getSize())
        print('Item on the list :')
        while currentNode is not None:
            print(currentNode.getData())
            currentNode = currentNode.next_ptr


def main():
    DLL = LinkList()
    DLL.addNode(6)
    DLL.addNode(2)
    DLL.addNode(3)
    DLL.addNode(8)
    DLL.addNode(4)

    DLL.printList()


if __name__ == "__main__":
    main()

File: linkedlist/SortedDoubleLinkedList/DoubleLinkedList.cpp
/* 
 * File:   DoubleLinedList.cpp
 * Author: tomvolek  
 * Insertion: O(1)
 * Search: O(n)
 * Deletion: O(n)
 * Fcuntion: Sorted Double Linked List, insertion() and printlist() 
 * Created on March 20, 2018, 11:59 PM
 */

#include <cstdlib>
#include <iostream>
#include <cstdlib>

using namespace std;

struct Node{
        int data;
         Node *next_ptr;
         Node *previous_ptr;
    } ;
    
class DoubleLinkedList {
      private: Node *head, *tail;
  
      // Constructor  
    public: DoubleLinkedList() {
        int value;
        head = NULL;
        tail = NULL;
    }
    
     void insert(int newdata){
     Node *newNode = new Node;
     newNode->data  = newdata ;
     newNode->next_ptr = NULL;
     newNode->previous_ptr = NULL; 
     
    if (head == NULL){
        head = newNode; 
        tail = newNode; 
        newNode = NULL; //free up the created newNode space 
        return;
        } 
    else {
        
        Node *currentNode = new Node;
        currentNode = head;
        while (currentNode != NULL  ){
            if (newNode->data <= currentNode->data){
                       // if (newNode->data < head->data){
                       if (currentNode->previous_ptr == NULL){
                           newNode->next_ptr = head;
                           head->previous_ptr = newNode;
                           cout << newNode->data << " Added to Head\n" ;
                           head = newNode;
                           return;
                       }
                       
                       //currentNode->previous_ptr = newNode; 
                       if (NULL != currentNode->previous_ptr)
                       {
                           currentNode->previous_ptr->next_ptr = newNode; 
                            newNode->next_ptr = currentNode;
                            newNode->previous_ptr = currentNode->previous_ptr; // needed ? 
                            cout << newNode->data << " Added to middle\n" ;
                            return; 
                       }

                   }
                   else { 
                       if (currentNode->next_ptr == NULL ) { //we have reached end of list, add to the end
                           currentNode->next_ptr = newNode; 
                           newNode->previous_ptr = currentNode;
                           tail = newNode;
                           cout << newNode->data << " Added to tail";
                           return ; 
                       }
                       else {
                           currentNode = currentNode->next_ptr ;
                         
                       }
                   }
            
        }
    
    
    }
            

    }  // End insert() 

    void printList(){
        Node *currentNode = head; 
           while (currentNode != NULL){
                cout << currentNode->data << " " ;
                currentNode = currentNode->next_ptr;
           }
           cout << "\n";
           
    } // End printList()

};  // End DoubleLinkedList 
    
 

int main(int argc, char** argv) {
    
           DoubleLinkedList DLL ;  //instantiate an empty  new Double linked list
           
            DLL.insert(6);
            DLL.printList();
            
            DLL.insert(3);
            DLL.printList();
            
            DLL.insert(5);
            DLL.printList();
            
            DLL.insert(2);
            DLL.printList();
            
            DLL.insert(7);
            DLL.printList();
            
            
            DLL.insert(8);
            DLL.printList();
            
            DLL.insert(10);
            DLL.printList();
            
            DLL.insert(7);
            DLL.printList();
            
            DLL.insert(7);
            DLL.printList();
            
            DLL.insert(5);
            DLL.printList();
    return 0;
}
